From 9c037bca4d42b94bf2b89201c1fc8ad068eb9a03 Mon Sep 17 00:00:00 2001
Message-Id: <9c037bca4d42b94bf2b89201c1fc8ad068eb9a03.1428007496.git.feitong.yi@intel.com>
In-Reply-To: <96b4927b026b1cf56e9298f537d7ca517a5902b4.1428007496.git.feitong.yi@intel.com>
References: <96b4927b026b1cf56e9298f537d7ca517a5902b4.1428007496.git.feitong.yi@intel.com>
From: sivakumar thulasimani <sivakumar.thulasimani@intel.com>
Date: Mon, 16 Mar 2015 17:25:57 +0530
Subject: [PATCH 5/6] FOR_UPSTREAM [VPG]: drm/i915: Fix for edp hpd loop

HPD is enabled for all displays by default, due to which eDP gets
hpd calls for VDD on/PPS on and VDD off/PPS off.

But i915 code to handle HPD for eDP turns on VDD for DPCD read and
then turns it off back again. so when the panel is put to sleep and
the edp panel is powered down it generates first HPD. The code to
handle this hpd turns on HPD and turns if off back again, which again
generates another HPD. thus becoming an infinite loop.

This patch disables HPD bits for eDP panel alone and this avoid the
scenario in the first place.

Issue:GMINL-6962
Change-Id: Ia619cb7b7c31724fea9ea3061ab29ecb735a340d
Signed-off-by: Sivakumar Thulasimani <sivakumar.thulasimani@intel.com>
---
 drivers/gpu/drm/i915/i915_irq.c |   18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index c73d9a9..2ae0daa 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -4933,11 +4933,21 @@ static void i915_hpd_irq_setup(struct drm_device *dev)
 	if (I915_HAS_HOTPLUG(dev)) {
 		hotplug_en = I915_READ(PORT_HOTPLUG_EN);
 		hotplug_en &= ~HOTPLUG_INT_EN_MASK;
-		/* Note HDMI and DP share hotplug bits */
-		/* enable bits are the same for all generations */
-		list_for_each_entry(intel_encoder, &mode_config->encoder_list, base.head)
-			if (dev_priv->hpd_stats[intel_encoder->hpd_pin].hpd_mark == HPD_ENABLED)
+
+		/*
+		 * Note HDMI and DP share hotplug bits
+		 * enable bits are the same for all generations
+		 * Also, avoid enabling HPD for eDP since it is not required
+		 */
+		list_for_each_entry(intel_encoder,
+			&mode_config->encoder_list, base.head) {
+			int hpd_status;
+			int port = intel_encoder->hpd_pin;
+			hpd_status = dev_priv->hpd_stats[port].hpd_mark;
+			if ((intel_encoder->type != INTEL_OUTPUT_EDP) &&
+				(hpd_status == HPD_ENABLED))
 				hotplug_en |= hpd_mask_i915[intel_encoder->hpd_pin];
+		}
 		/* Programming the CRT detection parameters tends
 		   to generate a spurious hotplug event about three
 		   seconds later.  So just do it once.
-- 
1.7.9.5

